src/Env.hs

```diff
@@ -0,0 +1,320 @@
+     return (AVar i)
+tParser' = do
+    t <- tParser
+    return (AT t)
+
+kParser, kTypeParser, tParser'' :: Parser K
+kParser = try kTypeParser <|> try tParser''
+kTypeParser = do
+     rword "type"
+     pos <- getPosition
+     return (Ktype (Type { typeName = "type", typePos = pos }))
+tParser'' = do
+    t <- tParser
+    return (KT t)
+
+emptyArgList :: Parser [Arg]
+emptyArgList = do
+  lparen
+  rparen
+  return []
+
+----------------------------------- Typechecker aux functions ------------------------------------------
+
+firsts :: [(a, b)] -> [a]
+firsts xs = [x | (x,_) <- xs]
+
+seconds :: [(a, b)] -> [b]
+seconds xs = [x | (_,x) <- xs]
+
+second :: (a, b) -> b
+second (a, b) = b
+
+checkAndGet :: String -> M.Map String v -> SourcePos -> Either String v
+checkAndGet k m pos = case M.lookup k m of
+  Nothing -> Left ("Error in " ++ (sourcePosPretty pos) ++ " : " ++ k ++ " Doesn't exist in the context \n")
+  Just v ->  Right v
+
+lookUpK :: VarEnv -> Arg -> K
+lookUpK e (AT  (TTypeVar t))  = (Ktype ((typeVarMap e) M.! t)) --(Ktype (Type {typeName = "type", typePos = typeVarPos t }))
+lookUpK e (AT  (TConstr t)) =
+        if (nameCons t) `elem` (declaredNames e)
+        then lookUpK e (AVar (VarConst (nameCons t)))
+        else (Ktype (Type { typeName = "type", typePos = constructorInvokerPos t }))
+lookUpK e (AVar v) = (KT ((varMap e) M.! v))
+
+getTypesOfArgs :: VarEnv -> [Arg] -> [K]
+getTypesOfArgs e args = map (lookUpK e) args
+
+updateEnv :: VarEnv -> (Y, K) -> VarEnv
+updateEnv e (TypeVarY y, Ktype t) = e { typeVarMap = M.insert y t $ typeVarMap e }
+updateEnv e (VarY y, KT t)        = e { varMap = M.insert y t $ varMap e }
+updateEnv e err                   = e { errors = (errors e) ++ "Problem in update: " ++ (show err) ++ "\n" }
+
+addName :: String -> VarEnv -> VarEnv
+addName a e = if (a `elem` (names e))
+              then e {errors = (errors e) ++ ("Name already exsist in the context \n")}
+              else e {names = a : (names e)}
+
+addDeclaredName :: String -> VarEnv -> VarEnv
+addDeclaredName a e = if (a `elem` (declaredNames e))
+                      then e { errors = (errors e) ++ ("Name already exsist in the context \n") }
+                      else e { declaredNames = a : (declaredNames e) }
+
+--------------------------------------- Env Data Types ---------------------------------------
+
+-- | Environment for the dsll semantic checker. As the 'check' function executes, it
+-- accumulate information such as symbol tables in the environment.
+
+-- | list of elements that might appear in the global context
+
+data Ttype = Ttype { yt :: Y,
+                     kt :: K }
+             deriving (Show, Eq, Typeable)
+
+data TypeConstructor = TypeConstructor { nametc :: String,
+                                         klstc  :: [K],
+                                         typtc  :: Type }
+                       deriving (Show, Eq, Typeable)
+
+data VarConstructor = VarConstructor { namevc :: String,
```

Comment:
these are *value* constructors, not variable constructors, and I would rename everything accordingly